### PR TITLE
@W-13198789 | Update Custom Orchestrator Sample

### PR DESCRIPTION
**Pull request by jbisaSF**
**Overview**

Updated the sample custom orchestrator to use:

- `BuyerActions` to determine which calculator to invoke
- `OptionalBuyerActionDetails` to potentially optimize each calculator that is invoked
- `postShipping` in order to place an order

**Testing Details**

- Checkout + Place Order succeeds with this custom orchestrator
- `checkout.connect.resources.CartSummaryServicePlaceOrderB2CTest.testPlaceOrderAsBuyerWithSampleCartSummaryOrchestrator` passes with this custom orchestrator

Screenshot 2023-07-28 at 12 28 42 PM

### CODE REVIEW

File: commerce/domain/orchestrators/classes/CartCalculateSample.cls [hunk 1]
```diff
@@ -3,26 +3,27 @@
...
         CartExtension.CartValidationOutputList cvoList  = cart.getCartValidationOutputs();
+        CartExtension.OptionalBuyerActionDetails optionalBuyerActionDetails = request.getOptionalBuyerActionDetails();
 
         // We might want to determine which calculator we call based on the state of the cart.
         // This block is executed when starting or modifying checkout.
         if (cartStatus == CartExtension.CartStatusEnum.CHECKOUT) {
             // Order matters. If the promotions calculator returns an error, we will finish execution early instead of
             // calling the other calculators
-            promotions(new CartExtension.CartCalculateCalculatorRequest(cart));
+            promotions(new CartExtension.CartCalculateCalculatorRequest(cart, optionalBuyerActionDetails));
```
Comment (pplat23) [inline commerce/domain/orchestrators/classes/CartCalculateSample.cls:20]: Can you just construct it once in one place above this if block and then pass it to every calculator?

**Review by gurpreetsainisalesforce — APPROVED** [review]

File: commerce/domain/orchestrators/classes/CartCalculateSample.cls [hunk 2]
```diff
@@ -3,26 +3,28 @@
...
-public class CartCalculateSample extends CartExtension.CartCalculate {
+global class CartCalculateSample extends CartExtension.CartCalculate {
 
     // Each orchestrator must override this calculate method.
-    public virtual override void calculate(CartExtension.CartCalculateOrchestratorRequest request) {
+    global virtual override void calculate(CartExtension.CartCalculateOrchestratorRequest request) {
         CartExtension.Cart cart = request.getCart();
         CartExtension.CartStatusEnum cartStatus = cart.getStatus();
         CartExtension.CartValidationOutputList cvoList  = cart.getCartValidationOutputs();
+        CartExtension.OptionalBuyerActionDetails optionalBuyerActionDetails = request.getOptionalBuyerActionDetails();
```
[the remark this answers is not in the document]
Comment (jbisaSF) [reply commerce/domain/orchestrators/classes/CartCalculateSample.cls]: Updated the sample to conditionally call calculators based on `BuyerActionDetails` and verified the changes with `testPlaceOrderAsBuyerWithSampleCartSummaryOrchestrator`. I'll need to make the same changes in Core as well.

File: commerce/domain/orchestrators/classes/CartCalculateSample.cls [hunk 3]
```diff
@@ -3,65 +3,67 @@
...
-
-                    if (hasErrorsInCVO(cvoList, CartExtension.CartValidationOutputTypeEnum.SHIPPING)) {
-                        return;
-                    }
-
-                    // This example takes the first available shipping option (CartDeliveryGroupMethod) and uses it as a
-                    // DeliveryMethod for the given CartDeliveryGroup.
-                    if (cartDeliveryGroup.getCartDeliveryGroupMethods().size() > 0) {
-                        String firstCartDeliveryGroupId = cartDeliveryGroup.getCartDeliveryGroupMethods().get(0).getDeliveryMethodId();
-                        cartDeliveryGroup.setDeliveryMethodId(firstCartDeliveryGroupId);
```
Comment (smahbubani99) [inline commerce/domain/orchestrators/classes/CartCalculateSample.cls:51]: do we no longer need to set a setDeliveryMethod to place an order?

Comment (jbisaSF) [reply commerce/domain/orchestrators/classes/CartCalculateSample.cls:51]: We do! This happens in `postShipping`: https://sourcegraph.soma.salesforce.com/perforce.soma.salesforce.com/app/main/core@HEAD/-/blob/commerce-checkout-impl/java/src/cartsummary/CalculatorService.java?L251 . See the `addShippingProductToCart` method in `ShippingCalculationServiceImpl.java`.

**Review by smahbubani99 — APPROVED** [review]

File: commerce/domain/orchestrators/classes/CartCalculateSample.cls [hunk 4]
```diff
@@ -3,65 +3,84 @@
...
+    global CartCalculateSample() {}
+
+    /**
+     * @description This constructor should only be exposed to customers in a test context
+     * @param apexExecutor Executor which can run the various calculators
+     */
+    global CartCalculateSample(final CartCalculateExecutorForUnitTest apexExecutor) {
+        super(apexExecutor);
+
+        if (!Test.isRunningTest()) {
```
Comment (pplat23) [inline commerce/domain/orchestrators/classes/CartCalculateSample.cls:20]: this is customer facing class. They don't have to write this in their classes.

File: commerce/domain/orchestrators/classes/CartCalculateSample.cls [hunk 5]
```diff
@@ -3,65 +3,84 @@
 // for orchestrators and then the extension must be linked to the webstore via the appropriate setup.
 
 // This class must extend the CartExtension.CartCalculate class
-public class CartCalculateSample extends CartExtension.CartCalculate {
+global class CartCalculateSample extends CartExtension.CartCalculate {
+
+    /**
+     * @description Default constructor
```
Comment (pplat23) [inline commerce/domain/orchestrators/classes/CartCalculateSample.cls:9]: Please add to comment that "All classes extending CartExtension.CartCalculate must have default constructor defined'

File: commerce/domain/orchestrators/classes/CartCalculateSample.cls (hunk 4 above)
Comment (jbisaSF) [reply commerce/domain/orchestrators/classes/CartCalculateSample.cls:20]: True, I can change the description to:

```This constructor should only be used in a test context```

**Review by pplat23 — APPROVED** [review]

**Pull request merged**